TODO/why-learn-functional-programming-in-javascript-composing-software.md

```diff
@@ -4,152 +4,151 @@
 
-JavaScript has the most important features needed for functional programming:
+JavaScript 有函数式编程所需要的最重要的特性：
 
-1. **First class functions:** The ability to use functions as data values: pass functions as arguments, return functions, and assign functions to variables and object properties. This property allows for higher order functions, which enable partial application, currying, and composition.
-2. **Anonymous functions and concise lambda syntax:**`x => x * 2` is a valid function expression in JavaScript. Concise lambdas make it easier to work with higher-order functions.
-3. **Closures:** A closure is the bundling of a function with its lexical environment. Closures are created at function creation time. When a function is defined inside another function, it has access to the variable bindings in the outer function, even after the outer function exits. Closures are how partial applications get their fixed arguments. A fixed argument is an argument bound in the closure scope of a returned function. In `add2(1)(2)`, `1` is a fixed argument in the function returned by `add2(1)`.
+1. **一级公民函数：**使用函数作为数据值的能力：用函数传参，返回函数，用函数做变量和对象属性。这个属性允许更高级别的函数，使偏函数应用、柯里化和组合成为可能。
+2. **匿名函数和简洁的 lambda 语法：**`x => x * 2` 是 JavaScript 中有效的函数表达式。简洁的 lambda 语法使得高阶函数变的简单。
+3. **闭包：**闭包是一个有着自己独立作用域的捆绑函数。闭包在函数被创建时被创建。当一个函数在另一个函数内部被创建，它可以访问外部函数的变量，即使在外部函数退出后。通过闭包偏函数应用可以获取内部固定参数。固定的参数时绑定在返回函数的作用域范围内的参数。在 `add2(1)(2)` 中，`1` 是 `add2(1)` 返回的函数中的固定参数。
 
-### What JavaScript is Missing
+### JavaScript 缺少了什么
 
-JavaScript is a multi-paradigm language, meaning that it supports programming in many different styles. Other styles supported by JavaScript include procedural (imperative) programming (like C), where functions represent a subroutine of instructions that can be called repeatedly for reuse and organization, object-oriented programming, where objects — not functions — are the primary building blocks, and of course, functional programming. The disadvantage of a multi-paradigm language is that imperative and object-oriented programming tend to imply that almost everything needs to be mutable.
+JavaScript 是多范式语言，意味着它支持多种风格的编程。其他被 JavaScript 支持的风格包括过程式（命令式）编程（比如 C），把函数看作可以被重复调用和组织的子程序指令；面向对象编程，对象— 而不是函数— 作为初始构造块；当然，还有函数式编程。多范式编程语言的劣性在于命令式和面向对象往往意味着所有东西都是可变的。
 
-Mutation is a change to data structure that happens in-place. For example:
+可变性指的是数据结构上的变化。比如：
+	
+	const foo = {
+	  bar: 'baz'
+	};
 
-    const foo = {
-      bar: 'baz'
-    };
+	foo.bar = 'qux'; // 改变
 
-    foo.bar = 'qux'; // mutation
+对象通常需要可变性以便于被方法更新值，在命令式的语言中，大部分的数据结构可变以便于数组和对象的高效操作。
 
-Objects usually need to be mutable so that their properties can be updated by methods. In imperative programming, most data structures are mutable to enable efficient in-place manipulation of objects and arrays.
+下面是一些函数式语言拥有但是 JavaScript 没有的特性：
 
-Here are some features that some functional languages have, that JavaScript does not have:
+1. **纯粹性：**在一些函数式语言中，纯粹性是强制的，有副作用的表达式是不被允许的。
```

Comment:
And more.
![image](https://cloud.githubusercontent.com/assets/8622362/24988163/c8ba547a-2036-11e7-9b5e-95fc657cb192.png)